Task/Factory.php

```diff
@@ -2,55 +2,110 @@
 
 namespace DigipolisGent\Domainator9k\CoreBundle\Task;
 
+use DigipolisGent\Domainator9k\CoreBundle\Ssh\Factory\SshShellFactoryInterface;
+use InvalidArgumentException;
+
 class Factory implements FactoryInterface
 {
-    protected static $map = array(
-        'provision.filesystem' => '\DigipolisGent\Domainator9k\CoreBundle\Task\Provision\Filesystem',
-        'provision.config_files' => '\DigipolisGent\Domainator9k\CoreBundle\Task\Provision\ConfigFiles',
-        'provision.cron' => '\DigipolisGent\Domainator9k\CoreBundle\Task\Provision\Cron',
-        'filesystem.create_directory' => '\DigipolisGent\Domainator9k\CoreBundle\Task\Filesystem\CreateDirectory',
-        'filesystem.create_file' => '\DigipolisGent\Domainator9k\CoreBundle\Task\Filesystem\CreateFile',
-        'filesystem.link' => '\DigipolisGent\Domainator9k\CoreBundle\Task\Filesystem\Link',
-        //'console.jenkins'               => '\DigipolisGent\Domainator9k\CoreBundle\Task\Console\Jenkins',
-        'console.cron' => '\DigipolisGent\Domainator9k\CoreBundle\Task\Console\Cron',
-    );
-
-    protected static $defaultOptions = array();
-
-    public static function setDefaultOptions(array $options = array())
+
+    protected $defaultOptions = array();
+
+    /**
+     * @var ShellFactoryInterface
+     */
+    protected $shellFactory;
+
+    /**
+     * Creates a new task factory.
+     *
+     * @param SshShellFactoryInterface $shellFactory
+     */
+    public function __construct(SshShellFactoryInterface $shellFactory)
     {
-        self::$defaultOptions = $options;
+        $this->shellFactory = $shellFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setDefaultOptions(array $options = array())
+    {
+        $this->defaultOptions = $options;
+    }
+
+    /**
+     * Add a new task definition.
+     *
+     * @param string $class
+     *
+     * @return $this
+     *
+     * @throws InvalidArgumentException
+     *     If the class does not implement the TaskInterface.
+     */
+    public function addTaskDefinition($class)
+    {
+        if (!is_subclass_of($class, TaskInterface::class)) {
+            throw new InvalidArgumentException(sprintf(
+                'Task %s does not implement %s.',
+                $class,
+                TaskInterface::class
+            ));
+        }
+        $this->map[call_user_func([$class, 'getName'])] = $class;
+
+        return $this;
     }
 
     /**
      * @param string $name
      * @param array  $options
      *
-     * @return AbstractTask
+     * @return AbstractSshTask
```

Comment:
Doesn't seem right, since you check for is_subclass_of($class, SshTaskInterface::class) below, shouldn't this remain AbstractTask type?